sasmodels/sasview_model.py

```diff
@@ -740,11 +745,20 @@ def _calculate_Iq(self, qx, qy=None):
                             magnetic=is_magnetic)
         lazy_results = getattr(calculator, 'results',
                                lambda: collections.OrderedDict())
-        #print("result", result)
 
         calculator.release()
         #self._model.release()
 
+        # Replace I and Q with there respective variable names in the code
+        expression = self.expression.replace("I", "result").replace("q", "qx").replace("Q", "qx")
+
+        # Calculate the expression string like a line of code
+        try:
+            result = eval(expression)
+        except NameError as e:
+            logging.error(f"Unknown symbol used in Y-axis Data, please only use I, Q and python mathematical operators"
+                          f"\n{e}")
+
```

Comment:
This is a plotting concern. Keep it in the plotter.

In particular, the result you return here will be used in fitting, so unless the data is in this transformed space your fits will not work. If your data is in a transformed space, then a pseudo resolution function similar to SESANS can be used to transform the SANS curve to the correct representation.

Also, `eval(...)` on arbitrary user expressions is a potential security risk. If you have a web backend performing calculations then this allows your users to run arbitrary code. It's okay if you restrict `eval` to the GUI since the user can already run arbitrary code on the machine.

Reply:
Hi paul, I originally tried to keep all the code in the plotter, but due to the workflow of the code calculate_Iq is called multiple times before the fitting and then throughout. Modifying the sasview_model.py was the only solution I could come up with.

Also the code is primarily used in plotting, but I think fundamentally it’s changing the model so maybe it’s a modelling concern? For example the core shell cylinder model is modified from:
`I(q,\alpha) = \frac{\text{scale}}{V_s} F^2(q,\alpha).sin(\alpha) + \text{background}`
to:
`I(q,\alpha) (q^2) = (\frac{\text{scale}}{V_s} F^2(q,\alpha).sin(\alpha) + \text{background}) (q^2).`

I can see what you mean with the eval expression. I can attempt to add a check to the user input code to ensure the string in the eval expression is a mathematical expression

Reply:
My solution to this sort of problem is to cache the calculated theory function rather than recalculating it each time it is needed.

Resolution, scaling and background should be applied to the theory function. Do you really want to calculate  scale (q² I(q)) ⊛ G(q, Δq) + background?

Note: convolution is a linear operator so the usual form so I(q) ⊛ G(q, Δq) = (scale P(q) + background)  ⊛ G(q, Δq) = scale (P(q) ⊛ G(q, Δq)) + background.

Regarding eval, you could write a parser but it is better if you don't have to. Do you really need the flexibility of arbitrary user expressions for plotting?

Reply:
Could you explain what you mean by “cache the calculated theory function”?

The user who requested this functionality mentioned that the exponential of q when plotting IQ verses Q depends on the experiment in question. The most common are IQ^4 and IQ^2, however the user also mentioned that IQ^3 will be useful to him.

My original idea was to have an “Q exponential” option, which would define the exponent of IQ, defaulting to 0 and accepting only floats as an input. But I thought this was confusing for the users and allowing the user to define their own expression gives them more flexibility, for example using I^2

Reply:
Cached theory as implemented in bumps curve fit:
https://github.com/bumps/bumps/blob/35902fae95031b199348f142db5a67fc10009c1c/bumps/curve.py#L237-L240
https://github.com/bumps/bumps/blob/35902fae95031b199348f142db5a67fc10009c1c/bumps/curve.py#L248-L254

"exponential" is the wrong term for power law.

You don't need the `eval` every time you do the transform, only when you define the transform. That is, you can say `transform = eval(f"lambda x, y: {transform_str}")` and pass the transform function to the plotter. I suggest `x` and `y` rather than `q` and `Iq` since not all plots are operating on `(q, Iq)`. I'm passing both `x` and `y` so that you can define, e.g., `y/x**3` for I(q)/q³.

Regarding open vs closed list of functions, that's up to the user community.

Reply:
I will try and implement the cached theory, but in the meantime I've replaced the eval command with a new function called safeEval, which should only allow the user to input mathematical operators, to alleviate the safety concerns